File: backend/services/git_service.py
from __future__ import annotations

import os
import uuid
from datetime import datetime, timezone

from git import Repo

from models.schemas import CommitInfo, RepoInfo


class GitService:
    """GitPython wrapper for managing Context Repos."""

    def __init__(self, data_dir: str):
        self.data_dir = os.path.abspath(data_dir)
        os.makedirs(self.data_dir, exist_ok=True)

    def _repo_path(self, repo_id: str) -> str:
        path = os.path.join(self.data_dir, repo_id)
        if not os.path.isdir(path):
            raise FileNotFoundError(f"Repo not found: {repo_id}")
        return path

    def _get_repo(self, repo_id: str) -> Repo:
        return Repo(self._repo_path(repo_id))

    # --- Repo CRUD ---

    def create_repo(self, name: str) -> RepoInfo:
        repo_id = uuid.uuid4().hex[:8]
        repo_path = os.path.join(self.data_dir, repo_id)
        os.makedirs(repo_path)

        repo = Repo.init(repo_path)

        # Store repo name in a metadata file
        meta_path = os.path.join(repo_path, ".contextgit.yaml")
        with open(meta_path, "w", encoding="utf-8") as f:
            f.write(f"name: {name}\ncreated: {datetime.now(timezone.utc).isoformat()}\n")

        repo.index.add([".contextgit.yaml"])
        repo.index.commit("Initial commit: create context repo")

        return RepoInfo(
            id=repo_id,
            name=name,
            current_branch="master",
            branches=["master"],
        )

    def list_repos(self) -> list[RepoInfo]:
        repos: list[RepoInfo] = []
        if not os.path.isdir(self.data_dir):
            return repos

        for entry in os.listdir(self.data_dir):
            full = os.path.join(self.data_dir, entry)
            if not os.path.isdir(full) or entry.startswith("."):
                continue
            git_dir = os.path.join(full, ".git")
            if not os.path.isdir(git_dir):
                continue

            try:
                repo = Repo(full)
                name = self._read_repo_name(full)
                branches = [h.name for h in repo.heads]
                current = repo.active_branch.name if not repo.head.is_detached else "HEAD"
                repos.append(RepoInfo(
                    id=entry,
                    name=name,
                    current_branch=current,
                    branches=branches,
                ))
            except Exception:
                continue

        return repos

    def get_repo_info(self, repo_id: str) -> RepoInfo:
        repo = self._get_repo(repo_id)
        name = self._read_repo_name(self._repo_path(repo_id))
        branches = [h.name for h in repo.heads]
        current = repo.active_branch.name if not repo.head.is_detached else "HEAD"
        return RepoInfo(
            id=repo_id,
            name=name,
            current_branch=current,
            branches=branches,
        )

    def delete_repo(self, repo_id: str) -> None:
        import shutil
        import stat
        path = self._repo_path(repo_id)

        def on_error(func, fpath, _exc_info):
            os.chmod(fpath, stat.S_IWRITE)
            func(fpath)

        shutil.rmtree(path, onexc=on_error)

    # --- Branch ---

    def list_branches(self, repo_id: str) -> list[str]:
        repo = self._get_repo(repo_id)
        return [h.name for h in repo.heads]

    def create_branch(self, repo_id: str, branch_name: str) -> list[str]:
        repo = self._get_repo(repo_id)
        repo.create_head(branch_name)
        return [h.name for h in repo.heads]

    def checkout(self, repo_id: str, branch_name: str) -> str:
        repo = self._get_repo(repo_id)
        repo.heads[branch_name].checkout()
        return branch_name

    # --- Commit ---

    def commit(self, repo_id: str, message: str) -> CommitInfo:
        repo = self._get_repo(repo_id)
        repo.index.add("*")
        c = repo.index.commit(message)
        return CommitInfo(
            hash=c.hexsha[:7],
            message=c.message,
            author=str(c.author),
            date=datetime.fromtimestamp(c.committed_date, tz=timezone.utc).isoformat(),
        )

    def get_log(self, repo_id: str, max_count: int = 50) -> list[CommitInfo]:
        repo = self._get_repo(repo_id)
        commits: list[CommitInfo] = []
        for c in repo.iter_commits(max_count=max_count):
            commits.append(CommitInfo(
                hash=c.hexsha[:7],
                message=c.message,
                author=str(c.author),
                date=datetime.fromtimestamp(c.committed_date, tz=timezone.utc).isoformat(),
            ))
        return commits

    def get_diff(self, repo_id: str, ref1: str = "HEAD~1", ref2: str = "HEAD") -> str:
        repo = self._get_repo(repo_id)
        try:
            return repo.git.diff(ref1, ref2)
        except Exception:
            return ""

    # --- File I/O ---

    def read_file(self, repo_id: str, filename: str) -> str | None:
        path = os.path.join(self._repo_path(repo_id), filename)
        if not os.path.isfile(path):
            return None
        with open(path, "r", encoding="utf-8") as f:
            return f.read()

    def write_file(self, repo_id: str, filename: str, content: str) -> None:
        path = os.path.join(self._repo_path(repo_id), filename)
        with open(path, "w", encoding="utf-8") as f:
            f.write(content)

    # --- Phase 2 stubs ---

    def diff_branches(self, repo_id: str, source: str, target: str) -> str:
        repo = self._get_repo(repo_id)
        return repo.git.diff(target, source)

    def merge_branch(self, repo_id: str, source: str, target: str) -> str:
        raise NotImplementedError("Merge will be implemented in Phase 2")

    # --- Helpers ---

    def _read_repo_name(self, repo_path: str) -> str:
        meta = os.path.join(repo_path, ".contextgit.yaml")
        if os.path.isfile(meta):
            with open(meta, "r", encoding="utf-8") as f:
                for line in f:
                    if line.startswith("name:"):
                        return line.split(":", 1)[1].strip()
        return os.path.basename(repo_path)

File: extension/src/services/git-service.ts
import git from "isomorphic-git";
import { getFs, deleteFs } from "../storage/fs";
import { db } from "../storage/db";
import type { RepoInfo, CommitInfo } from "../types/repo";
import { createTwoFilesPatch } from "diff";

const DIR = "/";
const AUTHOR = { name: "ContextGit", email: "user@contextgit" };
const META_FILE = ".contextgit.yaml";

export class GitService {
  // --- Repo CRUD ---

  async createRepo(name: string): Promise<RepoInfo> {
    const id = crypto.randomUUID().replace(/-/g, "").slice(0, 8);
    const fs = getFs(id);

    await git.init({ fs, dir: DIR, defaultBranch: "main" });

    const metaContent = `name: ${name}\ncreated: ${new Date().toISOString()}\n`;
    await fs.promises.writeFile(`${DIR}${META_FILE}`, metaContent, "utf8");

    await git.add({ fs, dir: DIR, filepath: META_FILE });
    await git.commit({
      fs,
      dir: DIR,
      message: "Initial commit: create context repo",
      author: AUTHOR,
    });

    await db.repos.add({ id, name, created: new Date().toISOString() });

    return { id, name, current_branch: "main", branches: ["main"] };
  }

  async listRepos(): Promise<RepoInfo[]> {
    const metas = await db.repos.toArray();
    const results: RepoInfo[] = [];

    for (const meta of metas) {
      try {
        const info = await this.getRepoInfo(meta.id);
        results.push(info);
      } catch {
        // Repo data corrupted or missing, skip
      }
    }
    return results;
  }

  async getRepoInfo(repoId: string): Promise<RepoInfo> {
    const fs = getFs(repoId);
    const meta = await db.repos.get(repoId);
    const name = meta?.name ?? repoId;

    const branches = await git.listBranches({ fs, dir: DIR });
    const current =
      (await git.currentBranch({ fs, dir: DIR })) ?? "main";

    return { id: repoId, name, current_branch: current, branches };
  }

  async deleteRepo(repoId: string): Promise<void> {
    deleteFs(repoId);
    await db.repos.delete(repoId);
  }

  // --- Branch ---

  async listBranches(repoId: string): Promise<string[]> {
    const fs = getFs(repoId);
    return git.listBranches({ fs, dir: DIR });
  }

  async createBranch(repoId: string, branchName: string): Promise<string[]> {
    const fs = getFs(repoId);
    await git.branch({ fs, dir: DIR, ref: branchName });
    return git.listBranches({ fs, dir: DIR });
  }

  async checkout(repoId: string, branchName: string): Promise<string> {
    const fs = getFs(repoId);
    await git.checkout({ fs, dir: DIR, ref: branchName });
    return branchName;
  }

  // --- Commit ---

  async commit(repoId: string, message: string): Promise<CommitInfo> {
    const fs = getFs(repoId);

    // Stage all known files
    const files = (await fs.promises.readdir(DIR)) as string[];
    for (const file of files) {
      if (file === ".git") continue;
      await git.add({ fs, dir: DIR, filepath: file });
    }

    const sha = await git.commit({
      fs,
      dir: DIR,
      message,
      author: AUTHOR,
    });

    const [entry] = await git.log({ fs, dir: DIR, depth: 1 });
    return {
      hash: sha.slice(0, 7),
      message: entry.commit.message,
      author: entry.commit.author.name,
      date: new Date(entry.commit.author.timestamp * 1000).toISOString(),
    };
  }

  async getLog(repoId: string, maxCount = 50): Promise<CommitInfo[]> {
    const fs = getFs(repoId);
    const entries = await git.log({ fs, dir: DIR, depth: maxCount });

    return entries.map((entry) => ({
      hash: entry.oid.slice(0, 7),
      message: entry.commit.message,
      author: entry.commit.author.name,
      date: new Date(entry.commit.author.timestamp * 1000).toISOString(),
    }));
  }

  async getDiff(
    repoId: string,
    ref1 = "HEAD~1",
    ref2 = "HEAD",
  ): Promise<string> {
    const fs = getFs(repoId);
    try {
      const content1 = await this.readFileAtRef(fs, ref1, "context.yaml");
      const content2 = await this.readFileAtRef(fs, ref2, "context.yaml");
      return createTwoFilesPatch(
        "context.yaml",
        "context.yaml",
        content1,
        content2,
      );
    } catch {
      return "";
    }
  }

  // --- File I/O ---

  async readFile(repoId: string, filename: string): Promise<string | null> {
    const fs = getFs(repoId);
    try {
      const content = await fs.promises.readFile(`${DIR}${filename}`, "utf8");
      return content as string;
    } catch {
      return null;
    }
  }

  async writeFile(
    repoId: string,
    filename: string,
    content: string,
  ): Promise<void> {
    const fs = getFs(repoId);
    await fs.promises.writeFile(`${DIR}${filename}`, content, "utf8");
  }

  // --- Directory ---

  async ensureDir(repoId: string, dirPath: string): Promise<void> {
    const fs = getFs(repoId);
    const parts = dirPath.split("/").filter(Boolean);
    let current = DIR;
    for (const part of parts) {
      current = current === "/" ? `/${part}` : `${current}/${part}`;
      try {
        await fs.promises.stat(current);
      } catch {
        await fs.promises.mkdir(current);
      }
    }
  }

  // --- Advanced commit: stage all files recursively ---

  async commitAll(repoId: string, message: string): Promise<CommitInfo> {
    const fs = getFs(repoId);
    await this.stageAll(fs, DIR, "");
    const sha = await git.commit({
      fs,
      dir: DIR,
      message,
      author: AUTHOR,
    });
    const [entry] = await git.log({ fs, dir: DIR, depth: 1 });
    return {
      hash: sha.slice(0, 7),
      message: entry.commit.message,
      author: entry.commit.author.name,
      date: new Date(entry.commit.author.timestamp * 1000).toISOString(),
    };
  }

  private async stageAll(
    fs: ReturnType<typeof getFs>,
    dir: string,
    prefix: string,
  ): Promise<void> {
    const entries = (await fs.promises.readdir(
      prefix ? `${dir}${prefix}` : dir,
    )) as string[];
    for (const entry of entries) {
      if (entry === ".git") continue;
      const relativePath = prefix ? `${prefix}/${entry}` : entry;
      const fullPath = `${dir}${relativePath}`;
      try {
        const stat = await fs.promises.stat(fullPath);
        if (stat.isDirectory()) {
          await this.stageAll(fs, dir, relativePath);
        } else {
          await git.add({ fs, dir, filepath: relativePath });
        }
      } catch {
        // Skip if stat fails
      }
    }
  }

  // --- Merge ---

  async merge(
    repoId: string,
    sourceBranch: string,
    targetBranch = "main",
  ): Promise<void> {
    const fs = getFs(repoId);
    await git.checkout({ fs, dir: DIR, ref: targetBranch });
    await git.merge({
      fs,
      dir: DIR,
      ours: targetBranch,
      theirs: sourceBranch,
      author: AUTHOR,
    });
  }

  // --- Branch diff ---

  async getBranchDiff(
    repoId: string,
    baseBranch: string,
    compareBranch: string,
  ): Promise<string> {
    const fs = getFs(repoId);
    try {
      // Get file list from compare branch
      const compareLog = await git.log({
        fs,
        dir: DIR,
        ref: compareBranch,
        depth: 1,
      });
      if (compareLog.length === 0) return "";

      const { tree: compareTree } = await git.readTree({
        fs,
        dir: DIR,
        oid: compareLog[0].oid,
      });

      const baseLog = await git.log({
        fs,
        dir: DIR,
        ref: baseBranch,
        depth: 1,
      });

      const diffs: string[] = [];
      for (const entry of compareTree) {
        if (entry.type !== "blob") continue;
        const newContent = await this.readBlobContent(fs, entry.oid);
        let oldContent = "";
        if (baseLog.length > 0) {
          oldContent = await this.readFileAtRef(fs, baseBranch, entry.path);
        }
        if (newContent !== oldContent) {
          diffs.push(
            createTwoFilesPatch(
              `a/${entry.path}`,
              `b/${entry.path}`,
              oldContent,
              newContent,
            ),
          );
        }
      }
      return diffs.join("\n");
    } catch {
      return "";
    }
  }

  // --- Read file at specific branch (without checkout) ---

  async readFileAtBranch(
    repoId: string,
    branch: string,
    filename: string,
  ): Promise<string | null> {
    const fs = getFs(repoId);
    try {
      return this.readFileAtRef(fs, branch, filename) || null;
    } catch {
      return null;
    }
  }

  // --- Helpers ---

  private async readBlobContent(
    fs: ReturnType<typeof getFs>,
    oid: string,
  ): Promise<string> {
    const { blob } = await git.readBlob({ fs, dir: DIR, oid });
    return new TextDecoder().decode(blob);
  }

  private async readFileAtRef(
    fs: ReturnType<typeof getFs>,
    ref: string,
    filepath: string,
  ): Promise<string> {
    // Resolve ref to oid
    let oid: string;
    if (ref.endsWith("~1")) {
      const baseRef = ref.replace("~1", "");
      const logs = await git.log({ fs, dir: DIR, ref: baseRef, depth: 2 });
      if (logs.length < 2) return "";
      oid = logs[1].oid;
    } else {
      const logs = await git.log({ fs, dir: DIR, ref, depth: 1 });
      if (logs.length === 0) return "";
      oid = logs[0].oid;
    }

    // Read the tree at this commit
    const { tree } = await git.readTree({ fs, dir: DIR, oid });
    const entry = tree.find((e) => e.path === filepath);
    if (!entry) return "";

    const { blob } = await git.readBlob({ fs, dir: DIR, oid: entry.oid });
    return new TextDecoder().decode(blob);
  }
}

export const gitService = new GitService();
